openedx/core/djangoapps/lang_pref/middleware.py

```diff
@@ -21,17 +26,50 @@ class LanguagePreferenceMiddleware(object):
     def process_request(self, request):
         """
         If a user's UserPreference contains a language preference, use the user's preference.
+        Save the current language preference cookie as the user's preferred language.
         """
-        languages = released_languages()
-        system_released_languages = [seq[0] for seq in languages]
+        cookie_lang = request.COOKIES.get(LANGUAGE_COOKIE, None)
+        if cookie_lang:
+            if request.user.is_authenticated():
+                set_user_preference(request.user, LANGUAGE_KEY, cookie_lang)
 
+            accept_header = request.META.get(LANGUAGE_HEADER, None)
+            if accept_header:
+                current_langs = parse_accept_lang_header(accept_header)
+                # Promote the cookie_lang over any language currently in the accept header
+                current_langs = [(lang, qvalue) for (lang, qvalue) in current_langs if lang != cookie_lang]
+                if current_langs:
+                    max_qvalue = max(qvalue for lang, qvalue in current_langs)
+                else:
+                    max_qvalue = 0
+                current_langs.append((cookie_lang, max_qvalue + 1))
+                accept_header = ",".join("{};q={}".format(lang, qvalue) for (lang, qvalue) in current_langs)
```

Comment:
Apparently, `q` technically maxes out at 1. So, we'll just prepend the cookie language at `q=1` at the front of the list, I guess.